src/mlpack/methods/neighbor_search/spill_search_impl.hpp

```diff
@@ -0,0 +1,251 @@
+    const MetricType metric) :
+    neighborSearch(naive, singleMode, epsilon, metric),
+    tau(tau),
+    rho(rho),
+    leafSize(leafSize)
+{
+  if (tau < 0)
+    throw std::invalid_argument("tau must be non-negative");
+  if (rho < 0 || rho > 1)
+    throw std::invalid_argument("rho must be in the range [0,1]");
+}
+
+// Clean memory.
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+~SpillSearch()
+{
+  /* Nothing to do */
+}
+
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+void SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+Train(const MatType& referenceSet)
+{
+  if (Naive())
+    neighborSearch.Train(referenceSet);
+  else
+  {
+    // Build reference tree with proper value for tau.
+    Tree* tree = new Tree(referenceSet, tau, leafSize, rho);
+    neighborSearch.Train(tree);
+    // Give the model ownership of the tree.
+    neighborSearch.treeOwner = true;
+  }
+}
+
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+void SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+Train(MatType&& referenceSetIn)
+{
+  if (Naive())
+    neighborSearch.Train(std::move(referenceSetIn));
+  else
+  {
+    // Build reference tree with proper value for tau.
+    Tree* tree = new Tree(std::move(referenceSetIn), tau, leafSize, rho);
+    neighborSearch.Train(tree);
+    // Give the model ownership of the tree.
+    neighborSearch.treeOwner = true;
+  }
+}
+
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+void SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+Train(Tree* referenceTree)
+{
+  neighborSearch.Train(referenceTree);
+}
+
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+void SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+Search(const MatType& querySet,
+       const size_t k,
+       arma::Mat<size_t>& neighbors,
+       arma::mat& distances)
+{
+  if (Naive() || SingleMode())
+    neighborSearch.Search(querySet, k, neighbors, distances);
+  else
+  {
+    // For Dual Tree Search on SpillTrees, the queryTree must be built with non
+    // overlapping (tau = 0).
+    Tree queryTree(querySet, 0 /* tau */, leafSize, rho);
```

Comment:
Ok, in this case, it is a lot of work to duplicate the entire class.  Instead I think we should use template specialization, but we can't specialize only a single member of a template class, so instead you'll need to write an auxiliary method kinda like:

```
template<typename TreeType>
TreeType* BuildQueryTree(arma::mat& dataset, boost::disable_if<IsSpillTree<TreeType>>::value*)
{
  return new TreeType(dataset);
}

template<typename TreeType>
TreeType* BuildQueryTree(const arma::mat& dataset, boost::enable_if<IsSpillTree<TreeType>>::value)
{
  return new TreeType(dataset, 0, leafSize, rho);
}
```

I've omitted a lot of details and there are probably syntax errors but the basic idea is to just use SFINAE in a standalone helper function.  I guess maybe it would have been easier for me to just say "check out `BuildTree()` at the top of neighbor_search_impl.hpp"...! :)